### PR TITLE
Retrieve and set tags properly

### PR DESCRIPTION
**Pull request by mattcui**
To address bug [109793518](https://www.pivotaltracker.com/story/show/109793518)

### CODE REVIEW

**Comment by mattcui** [comment]
@maximilien Passed travis, please review, thanks.

File: softlayer/vm/fs_agent_env_service.go [hunk 1]
```diff
@@ -2,11 +2,17 @@ package vm
...
 import (
 	"encoding/json"
+	"time"
 
 	bosherr "github.com/cloudfoundry/bosh-utils/errors"
 	boshlog "github.com/cloudfoundry/bosh-utils/logger"
 )
 
+const (
+	maxAttempts = 30
```
Comment (maximilien) [inline softlayer/vm/fs_agent_env_service.go]: Constants should be all caps

File: softlayer/vm/softlayer_vm.go [hunk 2]
```diff
@@ -382,15 +382,21 @@ func (vm SoftLayerVM) DetachDisk(disk bslcdisk.Disk) error {
 // Private methods
 func (vm SoftLayerVM) extractTagsFromVMMetadata(vmMetadata VMMetadata) ([]string, error) {
 	tags := []string{}
+	status := ""
 	for key, value := range vmMetadata {
-		if key == "tags" {
-			stringValue, ok := value.(string)
-			if !ok {
+		if key == "compiling" || key == "job" || key == "index" || key == "deployment" {
```
Comment (maximilien) [inline softlayer/vm/softlayer_vm.go]: Should we not have a default for this? What if bosh started passing something different... Maybe default to unknown and log the fact that the key is unknown as well.

File: softlayer/vm/softlayer_vm_test.go [hunk 3]
```diff
@@ -223,29 +223,37 @@ var _ = Describe("SoftLayerVM", func() {
...
-				err := vm.SetMetadata(metadata)
+				err = vm.SetMetadata(metadata)
 
 				Expect(err).ToNot(HaveOccurred())
 				Expect(softLayerClient.DoRawHttpRequestResponseCount).To(Equal(0))
 			})
 
 			It("at least one tag found", func() {
-				err := vm.SetMetadata(metadata)
+				metadataBytes := []byte(`{
```
Comment (maximilien) [inline softlayer/vm/softlayer_vm_test.go]: Let's be defensive and also have a test when no tags are found and we end up with unknown as mentioned above

File: softlayer/vm/fs_agent_env_service.go (hunk 1 above)
Comment (maximilien) [inline softlayer/vm/fs_agent_env_service.go]: I'll comment in @jianqiu PR

Comment (maximilien) [inline softlayer/vm/fs_agent_env_service.go]: Also, separate the PRs. Remove the retry change from @jianqiu from this PR. I asked him to do the same on his w.r.t. To this tag change. Thx

File: softlayer/vm/softlayer_vm.go (hunk 2 above)
Comment (mattcui) [inline softlayer/vm/softlayer_vm.go]: If the key is unknown, we don't set it as a VM tag, just ignore it, we only set meaningful info to tags.

**Comment by maximilien** [comment]
@mattcui is this ready for review?

/cc @jianqiu

File: softlayer/vm/softlayer_vm_test.go (hunk 3 above)
Comment (mattcui) [inline softlayer/vm/softlayer_vm_test.go]: We already cover this part of test:
https://github.com/mattcui/bosh-softlayer-cpi/blob/os_reload/softlayer/vm/softlayer_vm_test.go#L204

```go
		Context("no tags found in metadata", func() {
			BeforeEach(func() {
				metadataBytes := []byte(`{
				  "director": "fake-director-uuid",
				  "name": "fake-director"
				}`)

				metadata = bslvm.VMMetadata{}
				err := json.Unmarshal(metadataBytes, &metadata)
				Expect(err).ToNot(HaveOccurred())
			})

			It("does not set any tag values on the VM", func() {
				err := vm.SetMetadata(metadata)

				Expect(err).ToNot(HaveOccurred())
				Expect(softLayerClient.DoRawHttpRequestResponseCount).To(Equal(0))
			})
		})
```

**Comment by mattcui** [comment]
@maximilien I responded your comments, I removed the following testcase, since it will not happen unless Bosh has a bug. The value of tags we are trying to grab should always not be empty
```go
			It("the tags value is empty", func() {
				metadataBytes := []byte(`{
				  "director": "fake-director-uuid",
				  "fake1": "fake-deployment",
				  "fake2": "buildpack_python"
				}`)
				metadata = bslvm.VMMetadata{}
				err := json.Unmarshal(metadataBytes, &metadata)
				Expect(err).ToNot(HaveOccurred())

				err = vm.SetMetadata(metadata)

				Expect(err).ToNot(HaveOccurred())
				Expect(softLayerClient.DoRawHttpRequestResponseCount).To(Equal(0))
			})
```

**Comment by mattcui** [comment]
@maximilien wait a minute to review and merge, @jianqiu reported the integration test for set metadata failed, I am looking into the problem now.

**Comment by mattcui** [comment]
@maximilien It's ready to review now.

**Comment by maximilien** [comment]
OK merged and change merge commit message. Next time, please please @mattcui and @jianqiu please separate the PRs. Cheers :beers:

**Pull request merged**